[ci skip] Expand information on message options

schneems opened the pull request:
Link to the more detailed message documentation when using a message option in validations.

schneems committed:
[ci skip] Expand information on message options

Link to the more detailed message documentation when using a message option in validations.

maclover7 commented:
👍

sgrif commented:
This appears in more than just this example. Also are you sure the example is incorrect? `%{value}` is meant to be expanded. If it isn't this is a bug. I'm not sure if `%{ value }` is treated the same though (My guess would be that it isn't)

vipulnsward commented:
This should actually work- https://github.com/vipulnsward/rails/blob/20317f3d4df3ef0e876b5b213551f0d7d81d0c19/activerecord/test/cases/validations/i18n_generate_message_validation_test.rb#L27-L27

@schneems is there an example where we can see this fail on codetraige?

schneems commented:
I didn't commit it. if `%{value}` works and `%{ value }` does not. We should add an explicit note

```
The `%{value}` will be replaced with the invalid value provided to the model when the validation fails. 
```

To be more ruby-esque I would say that message should take a block and the block should receive the value as an argument.

sgrif commented:
This isn't Ruby though, it's i18n. It's how interpolation works with i18n, since we don't have an actual parser.

vipulnsward commented:
> To be more ruby-esque I would say that message should take a block and the block should receive the value as an argument.

We already allow to pass procs for messages.

I din't notice the extra space. I see no documentation over at https://github.com/svenfuchs/i18n or https://github.com/svenfuchs/rails-i18n for this. We can probably mention it, but this should be documented on https://github.com/svenfuchs/i18n instead.

schneems commented:
Turns out there is a whole section dedicated to the message option. I added a link to that section in all the examples we are using message. I also added a note about how exactly the replacement works (using I18n) and that spaces aren't allowed.

robin850 commented on guides/source/active_record_validations.md:
```diff
@@ -792,7 +797,8 @@ for each validation helper. The `:message` option accepts a `String` or `Proc`.
 A `String` `:message` value can optionally contain any/all of `%{value}`,
 `%{attribute}`, and `%{model}` which will be dynamically replaced when
-validation fails.
+validation fails. This replacement is done using the I18n gem, and the
```
Well I guess that's just Ruby, that's not up to the I18n gem:

~~~ruby
>> "%{ foo }" % { foo: "bar" }
KeyError: key{ foo } not found
from (pry):1:in `%`
>> "%{foo}" % { foo: "bar" }
=> "bar"
~~~

Maybe you can just say:

> The placeholders must match exactly, no spaces are allowed.

vipulnsward replied:
@robin850 i18n does its own interpolation. 
https://github.com/svenfuchs/i18n/blob/master/lib/i18n/interpolate/ruby.rb
https://github.com/svenfuchs/i18n/blob/master/lib/i18n/backend/interpolation_compiler.rb
